File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        ans = [0]
        buy, sell = 100000, -1

        for price in prices:
            if price < buy:
                if sell >= 0:
                    ans.append(sell - buy)
                    sell = -1
                buy = price
                continue
            if price > sell:
                sell = price

        ans.append(sell - buy)

        return max(ans)
                

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int profit  = 0;
        int buy = Integer.MAX_VALUE;

        for (int price : prices) {
            if (profit < (price - buy)) {
                profit = price - buy;
            }
            
            if (price < buy) {
                buy = price;
            }
        }

        return profit;
    }
}